ui/plugin-system/src/model/panels.ts

```diff
@@ -12,13 +12,26 @@
 // limitations under the License.
 
 import { UnknownSpec } from '@perses-dev/core';
-import { Plugin } from './plugin-base';
+import React from 'react';
+import { OptionsEditorTab } from '../components';
+import { OptionsEditorProps, Plugin } from './plugin-base';
 
+export type PanelOptionsEditorComponent<T> = Pick<OptionsEditorTab, 'id' | 'label'> & {
+  content: React.ComponentType<OptionsEditorProps<T>>;
+};
 /**
  * Plugin the provides custom visualizations inside of a Panel.
  */
 export interface PanelPlugin<Spec = UnknownSpec> extends Plugin<Spec> {
   PanelComponent: React.ComponentType<PanelProps<Spec>>;
+  /**
+   * React component for the content of "Query" tab
+   */
+  PanelQueryEditorComponent?: React.ComponentType<OptionsEditorProps<Spec>>;
+  /**
+   * React components for custom tabs
+   */
+  panelOptionsEditorComponents?: Array<PanelOptionsEditorComponent<Spec>>;
```

Comment:
Ok zooming out to bit to think about the schema changes...

So, the requirement we know we want to implement is that every plugin that is a panel type shows a `JSON` tab (PS I think the tabs look so good)

Currently, we require `OptionsEditorComponent` for every plugin (not just PanelPlugins).

If I'm following, it seems like we moved the "awareness" about the tabs out of the `OptionsEditorComponent` components, and up into the `PluginSpecEditor` so that we'd be able to always append the JSON tab to the end of the list of tabs.

And that makes sense to me -- we need to be able to append that JSON tab inside the framework code, instead of inside of the developer/custom code. 

So, I'm thinking... what if the only change we made to schema was that `Plugin` takes `optionsEditorComponents` instead of `OptionEditorComponent`

In this case, all plugins now can accept an array of editor components, instead of just one. So, we would have the same API for all plugins. 

Then, for each of the panel plugins we currently have (stat, time-series, gauge, markdown), we could implement both the SomePanelQueryEditorComponent, SomePanelOptionsEditorComponent, and we could include these in the array that we feed into `optionsEditorComponents`. 

In this case, we're not yet adding an expectation that every panel plugin has to have a query editor... I could be totally wrong, but it seems like it could be nice to leave that open, since we came to this wanting to make sure to always have a JSON tab, but not necessarily always have a Query tab. 

So overall change would be:

```
interface Plugin<Spec> {
  optionsEditorComponents: Array<React.ComponentType<OptionsEditorProps<Spec>> (or whatever TS works)
}
```

Reply:
Open to thoughts...

The main things making me think about this approach: 
- hm, wish we didn't have to have a different API for editing panel plugins vs. all other plugins
- do we definitely want to make PanelQueryEditorComponent a dedicated/named field in the PanelPlugin schema?

Reply:
> what if the only change we made to schema was that Plugin takes optionsEditorComponents instead of OptionEditorComponent

The only concern is that only panel plugin needs to be aware of `tabs`. Having tabs require panel plugin authors to also pass in a label and content for each option editor component. 

> wish we didn't have to have a different API for editing panel plugins vs. all other plugins

i think this is ok since each plugin type is very different. For example, `PrometheusDatasource` currently does not even have an option editor component 

> do we definitely want to make PanelQueryEditorComponent a dedicated/named field in the PanelPlugin schema?

I added this to enforce consistent ui/ux since having a query editor is very common for a panel plugin. `PanelQueryEditorComponent` will always be displayed under the `Query` tab, and `Query` tab will always be displayed first before other custom tabs.

Reply:
but yeah these are all valid points, we can discuss this with rest of the team at the next meeting!

Reply:
Nah, that makes sense to me -- I wasn't thinking about the tabs, and how we don't use tabs for other types of plugins (queries, datasource)

Yeah you're right, it's nice to be setting up the panel plugins for success -- many panels do have queries, guarantee that the query is always displayed in Query tab and Query tab always goes first, UI will look good ✨ 

So yeah my mind is changed, I think doing special stuff for the panel plugins is good / enabling us to be  more successful at building an awesome UI -- panel plugins have the most complex editing needs, and we want the editing experience that helps tame that complexity (nicely organized and consistent, which this does). I'm in ✅

Reply:
So there is another way of thinking about this (not necessarily for this PR):

Don't have a `PanelQueryEditorComponent` at all. Instead, the Panel plugin could specify which DataResponse formats it supports and that way we can automatically render the correct Query components. This would also help us de-couple query editing and the panel options (similar to the approach Grafana takes).
 
Grafana assumes every panel will also receive a DataFrame but gives you the ability to disable query editing with a declarative config option `skipDataQuery` if you don't use it. For example, see [Markdown panel config](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/text/plugin.json#L6) for example

Our interface might look something like this:

```ts
export interface PanelPlugin<Spec = UnknownSpec> extends Plugin<Spec> {
  PanelComponent: React.ComponentType<PanelProps<Spec>>;
  supportedQueryInterfaces?: Array<'TimeSeriesQuery' | 'LogQuery' | 'TraceQuery'>;
  panelOptionsEditorComponents?: Array<PanelOptionsEditorComponent<Spec>>;
}
```

So Plugins would look like this:

```ts
// No editor will show `supportedQueryInterfaces` is not present
const markdownPanel: PanelPlugin<MarkdownPanelOptions> = {
  PanelComponent: MarkdownPanel,
  panelOptionsEditorComponents: [{ label: 'Markdown', component: MarkdownOptionsEditor}],
}

// Only editors that support TimeSeriesQueries
const timeSeriesPanel: PanelPlugin<TimeSeriesPanelOptions> = {
  PanelComponent: TimeSeriesPanel,
  supportedQueryInterfaces: ['TimeSeriesQuery'],
  panelOptionsEditorComponents: [{ label: 'Time series', component: TimeSeriesOptionsEditor}],
}

// Only editors that support LogQueries
const logPanel: PanelPlugin<LogPanelOptions> = {
  PanelComponent: LogPanel,
  supportedQueryInterfaces: ['LogQuery'],
  panelOptionsEditorComponents: [{ label: 'Logs', component: LogOptionsEditor}],
}

// Will show editors that support 'TimeSeriesQuery', 'LogQuery', 'TraceQuery'
const tablePanel: PanelPlugin<TablePanelOptions> = {
  PanelComponent: TablePanel,
  supportedQueryInterfaces: ['TimeSeriesQuery', 'LogQuery', 'TraceQuery'],
  panelOptionsEditorComponents: [{ label: 'Table', component: TableOptionsEditor}],
}
```

Reply:
Are you thinking you'll pick up with the additional changes in a fast follow PR or that it's something we'll come back to farther in the future? If the latter, can these additional thoughts get captured in a github discussion or something else that will be a little more durable than a PR after this gets merged?

Reply:
i think i'll pick up the changes in a fast follow PR